### PR TITLE
Fix banding from linear RGB color space conversion

### PR DESCRIPTION
**Pull request by surma**
I discovered this weekend that our sRGB <-> linear RGB conversion introduces significant banding (which is enabled by default the second you start using the resize preprocessor).

<details>

banding

</details>

It’s fine for us to use a LUT from sRGB -> linear RGB as our set of input values is well-defined. Using a LUT for the reverse process, though, introduced significant banding. 

This PR removes the linear RGB -> sRGB LUT in favor of a function.

Additionally, to make linear RGB conversion, premultiplying and scaling as high-fidelity as possible I am now converting to `RGBAF32` images (`f32` per channel) if have at least one of premultiplying or linear RGB conversion is enabled.

### CODE REVIEW

**Review by jakearchibald — APPROVED** [review]

File: codecs/resize/src/lib.rs [hunk 1]
```diff
@@ -44,21 +47,24 @@ fn converter_funcs(with_space_conversion: bool) -> ((fn(u8) -> f32), (fn(f32) ->
...
-            |v, a| ((v as f32) * 255.0 / (a as f32)).clamp(0.0, 255.0),
-        )
+        (|v, a| v * a, |v, a| v / a)
     } else {
-        (|v, _a| v as u8, |v, _a| v as f32)
+        (|v, _a| v, |v, _a| v)
     }
 }
 
+fn vec_with_len<T: Clone>(len: usize, item: T) -> Vec<T> {
```
Comment (RReverser) [inline codecs/resize/src/lib.rs]: This exists in stdlib as a standard macro.

File: codecs/resize/src/lib.rs [hunk 2]
```diff
@@ -77,44 +83,64 @@ pub fn resize(
     let num_input_pixels = input_width * input_height;
     let num_output_pixels = output_width * output_height;
 
-    let (to_linear, to_color_space) = converter_funcs(color_space_conversion);
-    let (premultiplier, demultiplier) = alpha_multiplier_funcs(premultiply);
+    let mut output_image: Vec<u8> = vec_with_len(num_output_pixels * 4, 0);
```
Comment (RReverser) [inline codecs/resize/src/lib.rs]:
```suggestion
    let mut output_image = vec![0u8; num_output_pixels * 4];
```

File: codecs/resize/src/lib.rs [hunk 3]
```diff
@@ -77,44 +83,64 @@ pub fn resize(
...
+        for j in 0..3 {
+            preprocessed_input_image[4 * i + j] = premultiplier(
+                to_linear(input_image[4 * i + j]),
+                (input_image[4 * i + 3] as f32) / 255.0,
+            );
+            preprocessed_input_image[4 * i + 3] = (input_image[4 * i + 3] as f32) / 255.0;
         }
     }
 
+    let mut unprocessed_output_image: Vec<f32> = vec_with_len(num_output_pixels * 4, 0.0f32);
```
Comment (RReverser) [inline codecs/resize/src/lib.rs]:
```suggestion
    let mut unprocessed_output_image = vec![0.0f32; num_output_pixels * 4];
```

File: codecs/resize/src/lib.rs (hunk 1 above)
Comment (surma) [reply codecs/resize/src/lib.rs]: ... and that macro would be? :D

Comment (surma) [reply codecs/resize/src/lib.rs]: ah just saw below, lol. Great job, GitHub.

File: codecs/resize/src/lib.rs [hunk 4]
```diff
@@ -44,21 +47,24 @@ fn converter_funcs(with_space_conversion: bool) -> ((fn(u8) -> f32), (fn(f32) ->
...
+        (|v, _a| v, |v, _a| v)
     }
 }
 
+fn vec_with_len<T: Clone>(len: usize, item: T) -> Vec<T> {
+    let mut v: Vec<T> = Vec::with_capacity(len);
+    v.resize(len, item);
+    return v;
+}
+
```
Comment (surma) [inline codecs/resize/src/lib.rs]:
```suggestion
```

File: codecs/resize/src/lib.rs [hunk 5]
```diff
@@ -77,44 +83,64 @@ pub fn resize(
...
+    let mut preprocessed_input_image: Vec<f32> = Vec::with_capacity(input_image.len());
+    preprocessed_input_image.resize(input_image.len(), 0.0f32);
+    for i in 0..num_input_pixels {
+        for j in 0..3 {
+            preprocessed_input_image[4 * i + j] = premultiplier(
+                to_linear(input_image[4 * i + j]),
+                (input_image[4 * i + 3] as f32) / 255.0,
+            );
+            preprocessed_input_image[4 * i + 3] = (input_image[4 * i + 3] as f32) / 255.0;
```
Comment (RReverser) [inline codecs/resize/src/lib.rs]: This doesn't use `j`, is it meant to be happening inside of the inner `j` loop?

Comment (surma) [reply codecs/resize/src/lib.rs]: Nope. `j` loops over the RGB channel indices. This one sets the alpha channel.

Comment (surma) [reply codecs/resize/src/lib.rs]: oh, I just understood what you mean lol. Yes, gonna move it.

**Review by RReverser — APPROVED** [review]
LGTM minus nits (just rebuild Wasm before merging).

**Pull request merged**